crypto/hpke.c

```diff
@@ -0,0 +1,2730 @@
+/*
+ * Copyright 2019-2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/**
+ * @file
+ * An OpenSSL-based HPKE implementation of RFC9180
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+
+#include <openssl/ssl.h>
+#include <openssl/rand.h>
+#include <openssl/kdf.h>
+#include <openssl/evp.h>
+#include <openssl/params.h>
+#include <openssl/param_build.h>
+#include <openssl/core_names.h>
+
+/*
+ * If we're building standalone (from github.com/sftcd/happykey) then
+ * include the former, if building from github.com/sftcd/openssl) then
+ * the latter
+ */
+#ifdef HAPPYKEY
+#include "hpke.h"
+#else
+#include <openssl/hpke.h>
+#endif
+
+#ifdef TESTVECTORS
+#include "hpketv.h"
+#endif
+
+/*
+ * Define this if you want loads printing of intermediate
+ * cryptographic values
+ */
+#undef SUPERVERBOSE
```

Comment:
Would our trace API make more sense for this?

Reply:
as above,

Reply:
Sorry, a bit more on that - the SUPERVERBOSE stuff is really about interop testing and shouldn't be needed for ops reasons so I figure it's better to omit that code from the PR (while keeping it in https://github.com/sftcd/happykey) so I'd say there's no real need to use the trace API stuff for that level of detail.